langserver/diagnostics.go

```diff
@@ -71,18 +71,31 @@ func (h *LangHandler) publishDiagnostics(ctx context.Context, conn jsonrpc2.JSON
 	return nil
 }
 
-func updateCachedDiagnostics(cachedDiagnostics diagnostics, newDiagnostics diagnostics, files []string) diagnostics {
-	// add/update existing diagnostics
-	for file, diags := range newDiagnostics {
-		cachedDiagnostics[file] = diags
-	}
-
-	// remove all cached diagnostics for each files that is not present in `newDiagnostics`
-	// and add an entry to the output diagnostics for them (to clean the clients)
+func syncCachedDiagnostics(cachedDiagnostics diagnostics, newDiagnostics diagnostics, source string, files []string) diagnostics {
 	for _, file := range files {
-		if _, ok := newDiagnostics[file]; !ok {
-			if _, ok := cachedDiagnostics[file]; ok {
-				delete(cachedDiagnostics, file)
+		_, fileInCache := cachedDiagnostics[file]
+
+		//remove all of the diagnostics for the given source/file combinations and add the new diagnostics to the cache.
+		i := 0
+		for _, diag := range cachedDiagnostics[file] {
+			if diag.Source != source {
+				cachedDiagnostics[file][i] = diag
+				i++
+			}
+		}
+		cachedDiagnostics[file] = append(cachedDiagnostics[file][:i], newDiagnostics[file]...)
```

Comment:
If I read the old implementation correctly, `file` is not guaranteed to be in `newDiagnostics`. Otherwise we could just iterator over the `newDiagnostics` map and not pass in the `files` slice.

Reply:
That is correct. If there are no issues found in a file, it will not be in `newDiagnostics`. The `files` slice is included for the purpose of clearing out diagnostics that are no longer applicable.

Reply:
oh yeah I misread this. So if it is missing, it will return a `nil` slice which is safe to pass to the `...` operator. cool.